[elastic-agent] initial instrumentation

stuartnelson3 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

Add some initial tracing to elastic-agent. I basically just looked through the code and was looking at lowhanging fruit for wrapping http handlers and clients. I'd love to have some of the EA dev's point out places I've missed that I could add more tracing, and ways to trigger these traces to view them in the UI.

todo:
- more instrumenting
- tests
- how to configure apm output? currently apm-agent-go just using default values and sending to the apm-server integration.

## Why is it important?

This will help in troubleshooting issues as more users start running EA.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

build elastic-agent off this branch, start it under fleet, add the APM integration. the agent should appear as an apm service.

## Related issues

closes https://github.com/elastic/beats/issues/28895

## Use cases

troubleshooting mis-behaving elastic-agents

## Screenshots

![image](https://user-images.githubusercontent.com/1398104/142402500-60bdbdef-f16d-424f-aaf9-326a6d1a72a1.png)
(traces are coming from a metricbeat grabbing proxied data from apm-server)

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:
```diff
@@ -18,9 +18,20 @@ import (
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+		}
+	}()
```
this seemed like it would be interesting, but I seemingly haven't triggered a download yet.

axw replied:
If you delete the bundled apm-server artifact from the image, it should download it when the apm integration is assigned

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29031/detail/PR-29031/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29031/detail/PR-29031/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29031/detail/PR-29031/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29031/detail/PR-29031/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_29031.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-11-18T10:49:10.100Z&rangeTo=2021-11-18T11:09:10.100Z&transactionName=Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=ffe471b7d9445398b2279cb94b5aea4e&transactionId=f6912f41b04ed621)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats





* Duration: 21 min 41 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

michel-laterman commented:
Haven't reviewed anything yet, but this PR is exciting! It also covers some diagnostics work that's on my backlog.
I can make time to help out if needed

axw commented on x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:
```diff
@@ -20,7 +22,17 @@ import (
+	// TODO: span type?
+	span, ctx := apm.StartSpan(ctx, "Upgrader.downloadArtifact()", "custom")
+	defer func() {
+		span.End()
```
If you call this before capturing the error, the error's timestamp may be after the span ends. I'd reverse the order.

axw commented on x-pack/elastic-agent/pkg/agent/cmd/run.go:
```diff
@@ -135,21 +136,27 @@ func run(streams *cli.IOStreams, override cfgOverrider) error {
 
+	// TODO: How do we get the version?
+	version := release.Version()
+	tracer, err := apm.NewTracer(agentName, version)
```
We should call `apm.DefaultTracer.Close()` somewhere if we're not using it.

stuartnelson3 replied:
do you have a recommendation between using the default tracer, or a new one?

axw replied:
Creating a new one is appropriate since you're configuring it via code. The default one is appropriate when configuring externally with environment variables.

axw commented on x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:
```diff
@@ -20,7 +22,17 @@ import (
-func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI string) (string, error) {
+func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI string) (_ string, err error) {
+	// TODO: span type?
+	span, ctx := apm.StartSpan(ctx, "Upgrader.downloadArtifact()", "custom")
```
I think the span type should be "app.internal" (i.e span type = app, span.subtype = internal), according to https://github.com/elastic/apm/pull/513. @SylvainJuge is that correct?

axw reviewed (commented):
Looking good so far!

Not too sure about the span names. Typically I'd name after higher level operations, rather than method names. Maybe that's just my bias though.

stuartnelson3 commented:
That would be much appreciated! I'm sort of flying blind at the moment as I'm not super familiar with the codebase, so any help would be great.

ruflin commented:
This is really exciting! Thanks for taking this on.

One thing I'm curious in the context of diagnostics: Assuming someone is not shipping the apm data to ES, would it still be possible to fetch some of it as part of the diagnostics call? Does that even make sense? @axw @stuartnelson3

michel-laterman commented:
@ruflin, I'm not entirely sure how the apm client works, I think that integration with the diagnostics command should be done as a separate PR.
I don't think the apm client has an option to write to a local file,  in-memory buffer, etc. so we'll likely have to figure out how we want collect this information for diagnostics later.

axw commented:
> One thing I'm curious in the context of diagnostics: Assuming someone is not shipping the apm data to ES, would it still be possible to fetch some of it as part of the diagnostics call? Does that even make sense?

@ruflin not out of the box, but it would be possible to do something like this by implementing a custom transport. e.g. we have a transport which decodes events and stores them in memory for testing purposes: https://github.com/elastic/apm-agent-go/blob/master/transport/transporttest/recorder.go

simitt commented:
@ruflin would you mind creating a dedicated issue for this kind of conversation. I believe we can isolate it and don't think we should block making progress with initial instrumentation by this.

axw commented:
@ruflin please ping me on said issue and I'll reply there so the answers don't get lost.

ruflin commented:
@axw https://github.com/elastic/beats/issues/29125

lizozom commented:
Followed @stuartnelson3 's comment to this PR.

I'm coordinating the APM instrumentation of various Stack components and I have a couple of questions - 

 - Are you aware of the [good configs for APM issue](https://github.com/elastic/kibana/issues/117492)? Not sure these will be identical for `beats`, but there are some settings that are important to avoid performance implications or excessive data being collected.
 - How will APM collection from `beats` is going to be turned on \ off?
 - Where will the data be reported to?
 - Do you have an estimate on the data volume in question (From initial estimates APM data is pretty massive)?

I'd love to help coordinate the reporting and collection of this data and make sure we're all aligned, so hit me up if you want to talk about it.

stuartnelson3 commented:
>Are you aware of the good configs for APM issue? Not sure these will be identical for beats, but there are some settings that are important to avoid performance implications or excessive data being collected.

I was not aware! Our initial requirements are to be able to specify the environment, the apm-server url, and the secret token / api-key. The main goal (right now) is for elastic to gather telemetry data when troubleshooting customer deployments on an ad hoc basis, ie. data collection should only be temporary. Setting a sample rate is something we'd want if we're going to gather data for long periods of time, which we may do in the future.

>How will APM collection from beats is going to be turned on \ off?

This is flexible, I think. The customer will either turn it on themselves or request that we turn it on; all other configuration will be hidden to them. This could be done either in the fleet UI, or in a cloud UI.

>Where will the data be reported to?

One of the elastic-owned telemetry elasticsearch clusters

>Do you have an estimate on the data volume in question (From initial estimates APM data is pretty massive)?

At this point I don't have any estimates. The data is only going to be useful for a very limited time; I expect that once the cause of the failure is known the actual traces won't have much value and can be deleted.

lizozom commented:
> I was not aware! Our initial requirements are to be able to specify the environment, the apm-server url, and the secret token / api-key. The main goal (right now) is for elastic to gather telemetry data when troubleshooting customer deployments on an ad hoc basis, ie. data collection should only be temporary. Setting a sample rate is something we'd want if we're going to gather data for long periods of time, which we may do in the future.

That's an excellent use case and it seems like it aligns pretty well with https://github.com/elastic/dev/issues/1443.
The cloud SREs already have a set of monitoring clusters deployed per region, called [RLC for logs and RMC for metrics](https://github.com/elastic/cloud/issues/91181). Those clusters are already in use by SREs, they are monitored for performance and cost (?) and have decent ILM to reduce stored data size.

Since APM data is big and we want to avoid unnecessary network costs, we're contemplating to re-use them to collect regional APM data as well (for internal troubleshooting as well). (@lucasmoore understands this much better than myself, if you have any technical questions).

We have an issue for cloud to [propagate the right `serverUrl` per region](https://github.com/elastic/cloud/issues/91696) into Kibana \ Elasticsearch deployments. I think it's going to be interesting to think if this can work for `beats`. Do you think we could pass regional settings to `beats`? 

> How will APM collection from beats is going to be turned on \ off?

This is flexible, I think. The customer will either turn it on themselves or request that we turn it on; all other configuration will be hidden to them. This could be done either in the fleet UI, or in a cloud UI.

> Where will the data be reported to?

See my first comment

> At this point I don't have any estimates. The data is only going to be useful for a very limited time; I expect that once the cause of the failure is known the actual traces won't have much value and can be deleted.

I used this resource to guesstimate the data size for Kibana frontend data. For all Kibanas out there (cloud and on prem), I got an estimate of ~100GB just to track page loads and app navigations per day. So yeah, the data is pretty massive and it mostly depends on the amount of transactions you're planning to report and the depth of their stack. We can try to guesstimate the volumes for `beats` together.

michel-laterman commented on x-pack/elastic-agent/pkg/agent/cmd/run.go:
```diff
@@ -135,21 +142,33 @@ func run(streams *cli.IOStreams, override cfgOverrider) error {
 
 	statusCtrl := status.NewController(logger)
 
+	// TODO: How do we get the version?
```
`release.Version()` as you have below should be the version of the agent (unless you wanted a different version number?)

stuartnelson3 replied:
Nope, that sounds perfect

michel-laterman commented on x-pack/elastic-agent/pkg/agent/configuration/instrumentation.go:
```diff
@@ -0,0 +1,20 @@
+package configuration
+
+// InstrumentationConfig configures APM Tracing.
+type InstrumentationConfig struct {
```
I think we need to add tls config here as well

stuartnelson3 replied:
That's a good point. Internally, `apmtransport.NewHTTPTransport()` is checking environment variables to find TLS configs. While that works, it is leaving TLS config to a different method (environment variables vs. explicitly setting in the yml config).

Based on [this comment](https://github.com/elastic/fleet-server/pull/898#discussion_r753943397) in the fleet instrumentation PR, it sounds like they're not wanting to handle environment variables.

@marclop do you have an idea how you'd like to handle TLS configuration? I'm not sure if you considered it for the PR, since it's just a first iteration.

marclop replied:
I did not address that in my initial PR and I had overlooked that part. I'm happy to open a follow up with what we decide here.

michel-laterman commented on x-pack/elastic-agent/pkg/agent/configuration/settings.go:
```diff
@@ -15,12 +15,13 @@ import (
+	RetryConfig           *retry.Config                   `yaml:"retry" config:"retry" json:"retry"`
+	MonitoringConfig      *monitoringCfg.MonitoringConfig `yaml:"monitoring" config:"monitoring" json:"monitoring"`
+	LoggingConfig         *logger.Config                  `yaml:"logging,omitempty" config:"logging,omitempty" json:"logging,omitempty"`
+	InstrumentationConfig *InstrumentationConfig          `yaml:"instrumentation,omitempty" config:"instrumentation,omitempty" json:"instrumentation,omitempty"`
```
I know that only a few settings that appear under the `agent.*` settings are written to `fleet.yml`, such as the `agent.monitoring` settings.
If we want to keep apm settings under `agent.instrumentation` we need to make sure that they can be read/written to `fleet.yml` as well.

Do we want to keep them under `agent.instrumentation` or move it under `agent.monitoring.`?
Another question is why `instrumentation` and not `apm` (the [docs](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html) seem to use `apm`)?

stuartnelson3 replied:
I was trying to stay consistent with the fleet instrumentation PR, which has already [been merged](https://github.com/elastic/fleet-server/pull/898/files#diff-7a93ad3d35c4a6f46f0b58e67a5c63503a0a5eefcc7fe48d15147e7331a107f7). We didn't have a formal discussion about the naming, though. If nesting under `agent.instrumentation` presents problems (how difficult is it to make it read/writable to fleet.yml?), we could change both this and the other PR.

I personally don't want to nest it under `agent.monitoring` just because it seems like it's about configuring exposing internally collected metrics, rather than pushing traces somewhere. This might just be an arbitrary separation my brain is making and to others it makes perfect sense to be nested under `monitoring`.

@marclop @ruflin @simitt @axw do you have an opinion on the nesting and naming?

axw replied:
Maybe a little bit of hysterical raisins. The term "instrumentation" was originally used because it originated in apm-server's config. i.e. we had

```yaml
apm-server:
  instrumentation:
    enabled: true
```

Meaning apm-server's own instrumentation is enabled. We thought it would be rather confusing to name it `apm-server.apm`. Later we moved the config to libbeat, so it would just work for other Beats too. At this point we didn't stop to rethink the config name. TBH I don't love "instrumentation" -- sounds almost as meaningful as "stuff", but on the other hand it _is_ a collection of different things: metrics, traces, errors...

I don't have a good idea of what it should be though. `agent.monitoring` as a heading sounds OK. @ruflin?

marclop replied:
I didn't stop to rethink the names, but it's something that crossed my mind, in the end I didn't pursue it further since it seems that `libbeat` uses the `instrumentation` term and that is the current working configuration setting, on the other I don't know what other name would be fitting, we're shipping more than traces, but `agent.tracing` seems to be a name that seems fitting, even if we ship more than traces?

ruflin replied:
@michel-laterman The `fleet.yml` part you refer to is the one that allows a user to have a managed elastic-agent with some local configs? If yes, it is something I don't like too much. It is not that we should not allow a user for some configs to have a local setting too but the way we implemented it I think adds more complexity than needed. We likely should take a step back here in general and discuss our approach forward before we add more settings to it and put us in a corner.

For the naming I recommend to disonnect from what we have in apm-server and fleet-server. If we would start from scratch, what would we name it in Elastic-Agent? We can still do this even if apm-server and fleet-server have different namings as we have a translation layer in place and fleet-server could change configs at any time as long as Elastic-Agent keeps coping with it, as it is never run standalone.

axw replied:
What about something like this?

 - introduce `agent.monitoring.traces: true|false` (defaults to false)
 - introduce `agent.monitoring.apm_server.{hosts,api_key,secret_token}`:

e.g.

```yaml
agent.monitoring:
    logs: true
    metrics: true
    traces: true # enables tracing with the APM Go agent, uses apm_server for tracer's transport
    apm_server:
        hosts:
          - https://apm-server.one.com:8200
          - https://apm-server.two.com:8200
        secret_token: a_secret_token
        api_key: or_an_api_key
```

That leaves room, in the future, for sending metrics via APM Server (which may be part of a future architecture).

simitt replied:
IMO it generally would perfectly fit under `agent.monitoring` as it is complementary to metrics and logs collection. _Maybe_ an even better name would be `agent.observability.apm`? Just throwing in ideas here. 

@ruflin if this PR should make it into 8.0, we need to know how to move forward with adding the config options for now. I suggest to document and label the addition of apm as `beta`, so that we do not lock us in to the config options if we are not certain they stick around like this in the future. I'd expect this to be mainly used internally for now for troubleshooting. 
Ideally we align the fleet-server config options after a decision has made on this PR.

simitt commented on x-pack/elastic-agent/pkg/agent/application/fleet_server_bootstrap.go:
```diff
@@ -133,6 +136,7 @@ func (b *FleetServerBootstrap) Start() error {
 	if err := b.srv.Start(); err != nil {
 		return err
 	}
+	// TODO: Does it make sense to pass in a ctx here?
```
Can you either clarify the TODOs and remove them or add a bit more detail, your github handle and/or a link to a follow up task?

stuartnelson3 replied:
:+1: i should start writing "QUESTION" since that's what they really are, they just get highlighed when I use `todo` in vim so i see them :sweat_smile: i've cleaned it up a bit since talking with @michel-laterman last week.

simitt reviewed (commented):
Generally this looks good; left a comment around the config options discussion.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

stuartnelson3 commented:
@michel-laterman I think this is ready for review. It seems like we may still be working out the configuration naming, but I don't think anything else will change

stuartnelson3 commented:
Could someone take a look at the E2E tests? The failures look like they're more to do with the infrastructure rather than the changes in this PR

simitt commented:
@ruflin can you chim in regarding the config options and if this should make it into `8.0` with the proposed options.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/gateway/fleet/fleet_gateway.go:
```diff
@@ -175,7 +175,8 @@ func (f *fleetGateway) worker() {
 			var errMsg string
-			if err := f.dispatcher.Dispatch(f.acker, actions...); err != nil {
+			// TODO: Is there a purpose in using f.bgContext?
+			if err := f.dispatcher.Dispatch(context.Background(), f.acker, actions...); err != nil {
```
context: `fleetGateway` has a `bgContext` struct member

ph replied:
I think we could refactor it to remove it and just pass it on call.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/pipeline/stream/operator_stream.go:
```diff
@@ -44,8 +48,17 @@ func (b *operatorStream) Specs() map[string]program.Spec {
-func (b *operatorStream) Execute(cfg configrequest.Request) error {
-	return b.configHandler.HandleConfig(cfg)
+func (b *operatorStream) Execute(ctx context.Context, cfg configrequest.Request) (err error) {
+	// TODO: double check `route` as name
```
`*operatorStream.Execute()` is only being called by `*router.Route`, so I'm wondering if that would be the right span name, or something else.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/operation/operator.go:
```diff
@@ -155,13 +157,20 @@ func (o *Operator) Close() error {
-		err = filterContextCancelled(err)
+		if err != nil {
+			apm.CaptureError(ctx, err).Send()
+			// TODO: do we want to report `context.Canceled`?
```
I assume that for tracing purposes this would be helpful, but let me know if you'd rather have it filtered out.

stuartnelson3 replied:
If the errors are being wrapped then it would, but I'm not familiar enough with the codebase to know how the errors are being handled. The noise was also my main concern.

stuartnelson3 replied:
Cool, I'll remove it for now

lykkin reviewed (commented):
exciting changes! a few questions/comments for my own edification, i'm new to the world of tracing in go and how we do things in elastic apm.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:
```diff
@@ -20,7 +22,14 @@ import (
 
-func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI string) (string, error) {
+func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI string) (_ string, err error) {
+	span, ctx := apm.StartSpan(ctx, "upgrade", "app.internal")
```
I think the url / filepath passed into the fn will be included in child spans, but it might be nice to include `sourceURI` as a [label on the span](https://pkg.go.dev/go.elastic.co/apm#SpanContext.SetLabel). I'm not sure where best to do this, but maybe up one level, in `*Upgrader.Upgrade()`?

@axw do you have an opinion about @lykkin's question?

axw replied:
```suggestion
	span, ctx := apm.StartSpan(ctx, "downloadArtifact", "app.internal")
```

Might be a more informative span name?

@lykkin yes as Stuart points out there's an API for attaching arbitrary labels (strings, booleans, or numbers). These are stored under `labels.<name>` in the resulting span document.

For this particular case I guess the `fetcher.Download` call down below will issue an HTTP request? If so, we should probably instrument that method's internal HTTP client with `apmhttp.WrapClient`. That way we'll see an HTTP client span, and that will have the requested URL and response status code. @stuartnelson3 are we already doing that?

stuartnelson3 replied:
Correct, the http client is created [here](https://github.com/elastic/beats/blob/master/x-pack/elastic-agent/pkg/artifact/download/http/downloader.go#L41-L43), which includes wrapping the roundtripper

stuartnelson3 commented on x-pack/elastic-agent/pkg/fleetapi/client/client.go:
```diff
@@ -36,7 +38,7 @@ type Sender interface {
 var baseRoundTrippers = func(rt http.RoundTripper) (http.RoundTripper, error) {
 	rt = NewFleetUserAgentRoundTripper(rt, release.Version())
-	return rt, nil
+	return apmhttp.WrapRoundTripper(rt), nil
```
right, this will propagate the trace with the correct tracing headers :) I haven't tried it yet, actually, but [fleet-server has been instrumented](https://github.com/elastic/fleet-server/pull/898) so it should work with a build based off this branch

ruflin commented:
I like the suggestion from @simitt to define it as beta or likely even as experimental in 8.0 so we keep us all the options open to rename it afterwards. Today I had a partially related discussion with @narph that we need to have a more detailed look at the config files and what is inside again and come up with "rules" on what should be where. 

Based on the above, lets pick the config that fits best right now, make it experimental but push in 8.1 / 8.2 to come up with the final name and make it GA.

simitt commented:
Any update on this? It seems it hasn't made it into `8.0` as `beta` after all. Can we ensure to get this in to `8.1`?

stuartnelson3 commented:
Yeah, this didn't make it into the 8.0 release as we were still trying to figure out how to handle the configuration.

@ruflin's comment is still relevant:

>Based on the above, lets pick the config that fits best right now, make it experimental but push in 8.1 / 8.2 to come up with the final name and make it GA.

Right now, it's nested under `monitoring.apm`, which was given as a suggestion in [this comment](https://github.com/elastic/beats/pull/29031#discussion_r762885909).

There are a couple questions still in the PR that haven't been answered, but other than those, I believe this is "ready".

ruflin commented:
@lykkin @michel-laterman Could one of you help to get this over the line? @jlind23 We should plan this in for 8.1.

@ph Maybe an interesting PR to get up to speed on the code again? ;-)

ph commented on x-pack/elastic-agent/pkg/agent/application/fleet_server_bootstrap.go:
```diff
@@ -7,6 +7,8 @@ package application
 import (
 	"context"
 
+	"go.elastic.co/apm"
```
Interesting I didn't know about go.elastic.co, when the restructuring of the repository is complete we should take a look a it. @kvch @cmacknz @ruflin

kvch replied:
TBH, I am not a huge fan of vanity imports. It looks great, but it is just vanity... :)

ph commented:
@stuartnelson3 This looks good to me as a first PR to add the feature, I haven't tested yet the behavior with Kibana.

Can you rebase your code I will try it out?

One question I have on the usage, I have been often using decorator to add instrumentation to the existing type, instead of coupling the instrumentation and the actual business logic. Is this a pattern that people users with APM or it would introduce false information in our tracing? I raising that because we often instrument public boundaries of packages. Note I understand that not all the types have a public interface which would increase the complexity of this PR, I am asking this as a preparation of v2 architecture and some of the things we were discussing for taping into pipelines.

stuartnelson3 commented on x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:
```diff
@@ -74,7 +76,15 @@ func (ad *ActionDispatcher) key(a fleetapi.Action) string {
+		}
+		span.End()
+	}()
+
```
there isn't one existing as of now but I could add it? for me personally it's not a big deal, and without knowing how code in the `defer` func might change between different function calls, I decided not to extract to a function yet.

stuartnelson3 commented:
>One question I have on the usage, I have been often using decorator to add instrumentation to the existing type, instead of coupling the instrumentation and the actual business logic. Is this a pattern that people users with APM or it would introduce false information in our tracing? I raising that because we often instrument public boundaries of packages. Note I understand that not all the types have a public interface which would increase the complexity of this PR, I am asking this as a preparation of v2 architecture and some of the things we were discussing for taping into pipelines.

within library code (like the modules in https://github.com/elastic/apm-agent-go) we use the decorator pattern to simplify adoption for our users, but for application code (like the apm-server) we typically add the instrumentation directly.

stuartnelson3 commented:
@ph i've addressed some comments and written some follow-up questions, can you take a look when you have a chance?

also, I'm not sure what the e2e test failure is, it's apparently something with kubernetes (`kubernetes_autodiscover_elastic-agent`). I merged in latest master and will see what happens...

ph reviewed (approved):
This looks good to me, @stuartnelson3 @blakerouse can you have an eye on it?

stuartnelson3 commented:
any idea what's up with the packaging failures?
```
[2022-01-25T19:23:45.582Z] ERRO[2022-01-25T19:23:45Z] Could not download the binary for the agent   arch=x86_64 artifact=elastic-agent error="GET request failed with 404" extension=tar.gz os=linux version=8.1.0-dbc834fd-SNAPSHOT
[2022-01-25T19:23:45.582Z] ERRO[2022-01-25T19:23:45Z] Could not download the binary for the agent   arch=x86_64 artifact=elastic-agent error="GET request failed with 404" extension=tar.gz os=linux version=8.1.0-dbc834fd-SNAPSHOT
```

blakerouse reviewed (approved):
Overall the change looks very mechanical, which is good to see. Made it easier to review such a large branch.

I reviewed the `initTracer` as that is probably the most important part. Looks good and provides unit test coverage for it.

stuartnelson3 commented:
/test

stuartnelson3 commented:
/test

stuartnelson3 commented:
/test

ruflin commented:
/test

stuartnelson3 commented:
/test

mdelapenya commented:
e2e are failing but not caused by this PR. We @elastic/observablt-robots are working on the permanent fix (very probably here elastic/e2e-testing#2064), so I'd say this PR is fine

👍
